lib/gc/Transforms/Microkernel/EarlyDispatchMicrokernel.cpp

```diff
@@ -0,0 +1,201 @@
+//===-- EarlyDispatchMicrokernel.cpp - Dispatch before runtime --*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include <sstream>
+
+#include "gc/Transforms/Microkernel/BrgemmRuntimeUtils.h"
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h"
+#include "gc/Transforms/Utils/ValueUtils.h"
+#include "oneapi/dnnl/dnnl_types.h"
+
+namespace mlir::microkernel {
+#define GEN_PASS_DEF_EARLYDISPATCHMICROKERNEL
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h.inc"
+
+#define DEBUG_TYPE "early-dispatch-microkernel"
+
+static FailureOr<std::string>
+createGlobalKernelHandleName(RewriterBase &rewriter,
+                             microkernel::BrgemmDispatchOp op) {
+  // TODO(haixin): Add runtime backend type to global name
+  std::stringstream ss;
+  ss << "g_dispatched_microkernel_brgemm";
+
+  auto flags = op.getFlagsAttr();
+  for (auto flag : flags) {
+    auto brgemmFlag = dyn_cast_or_null<microkernel::BrgemmFlagsAttr>(flag);
+    if (!brgemmFlag)
+      return failure();
+    if (brgemmFlag.getValue() == BrgemmFlags::LIST)
+      return failure();
+    if (brgemmFlag.getValue() == BrgemmFlags::BETA_0)
+      ss << "_init";
+  }
+
+  // M, N, K, LDA, LDB, LDC, stride_a, stride_b
+  // they are in the same order with BrgemmDispatchOp inputs
+  ArrayRef<int64_t> inputs = op.getInputsAttr().asArrayRef();
+  for (auto input : inputs) {
+    ss << "_" << input;
+  }
+
+  // dtypeA, dtypeB
+  auto dtypes = op.getDataType();
+  if (dtypes.size() != 2)
+    return failure();
+  ss << "_" << getDnnlDataTypeVal(rewriter, dtypes[0]);
+  ss << "_" << getDnnlDataTypeVal(rewriter, dtypes[1]);
+
+  return ss.str();
+}
+
+// get or create global kernel handle with initializer, identified by
+// `kernelName`
+static FailureOr<LLVM::GlobalOp>
+getOrCreateGlobalKernelHandle(RewriterBase &rewriter, ModuleOp module,
+                              const std::string &kernelName,
+                              microkernel::BrgemmDispatchOp op) {
+  // Create the global at the entry of the module
+  LLVM::GlobalOp global;
+  if (!(global = module.lookupSymbol<LLVM::GlobalOp>(kernelName))) {
```

Comment:
Shall we avoid using the return value of C++ assignment, `=`? Maybe we can refactor as
```
LLVM::GlobalOp global = module.lookupSymbol<LLVM::GlobalOp>(kernelName);
if(!global) {
```

Reply:
Refactored, also return early if `global` is valid to avoid large segment of codes wrapped in `if` branch.